---
title: "06_analysis_CARNIVAL_results"
author: "Rosa Hernansaiz-Ballesteros"
date: "14/08/2020"
output: 
  github_document: default
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### License Info

This program is free software: you can redistribute it and/or modify it under 
the terms of the GNU General Public License as published by the Free Software 
Foundation, either version 3 of the License, or (at your option) any later 
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR 
A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Please check http://www.gnu.org/licenses/.

## Introduction
This is the final part of our series of transcriptomics tutorials. In previous parts, we normalised RNASeq data for differential analysis, and used the differential analysis results in **PROGENy** and **DOROTHea**. In the previous tutorial, we demonstrated **CARNIVAL** with TF activities from **DOROTHea**, a prior knowledge network from **Omnipath**, and weights from **PROGENy**.

## CARNIVAL output

CARNIVAL (CAusal Reasoning for Network identification using Integer VALue programming) 
returns a list with 4 elements: 

* weightedSIF. Character matrix. Contains a summary of _sifAll_, 
where *Weight* indicates the number of networks that present that interaction.

* nodesAttributes. Character matrix. Contains a summary of _attributesAll_, 
where all columns ending in *Act* indicate the number of networks that present that node in
each of the states (Zero = doesn't appear in the network, Up = upregulated, Down = downregulated, Avg = average regulation).
The column *NodeType* indicates if the node is a TF's (T) or a perturbed one (S).

* sifAll. List of character matrices. All networks that have been calculated.

* attributesAll. List of character matrices.

NOTE: This is valid for current CARNIVAL version (1.0.0)

## CARNIVAL interpretation

The summary files ( _sifAll_ and _attributesAll_ ) can be directly used to identify causal interactions between the 
perturbed nodes and the selected Transcription Factors. 
In addition to extracting direct information from the network, 
we can run different downstream analysis based on the necessities of each project.

Here are described some of the downstream approaches that we have used:

+ Pathway enrichment analysis
+ Network comparisons

# Get ready

```{r libraries, message=FALSE}
library(readr)
library(piano)
library(dplyr)
library(ggplot2)
library(tibble)
library(tidyr)
library(dplyr)
library(scales)
library(plyr)
library(GSEABase)
library(network)
library(reshape2)
library(cowplot)
library(pheatmap)
library(ggraph)
library(tidygraph)
if (!require(RNAscripts)) {
  devtools::install("../scripts/RNAscripts", upgrade = "never", quiet = TRUE)
}
library(RNAscripts)
# set working directory
# setwd("~/Projects/transcriptutorial/scripts")
# source("./support_networks.r")
## We also load the support functions
if (exists("snakemake")) {
  carnival_sample_result <- snakemake@input[["carnival_sample_result"]]
  tt_basepath <- snakemake@params[["tutorial_source_path"]]

  names(carnival_sample_result) <- snakemake@params[["sample_names"]]
  stopifnot(length(carnival_sample_result) > 0)
  workflow_config <- snakemake@config

  sample_table <- readr::read_tsv(workflow_config$samples)
  organism <- workflow_config$organism
} else {
  BASE_DIR <- "/omics/odcf/analysis/OE0228_projects/VascularAging/rna_sequencing/tec_aplnrKO/results/carnival/sample_resolution"
  workflow_config <- yaml::read_yaml("../../../configs/tec_aplnrKO.yaml")
  sample_table <- readr::read_tsv("../../../data/tumor_vs_ec/Aplnr_KO.tsv")
  sample_ids <- sample_table %>% pull(sample)
  carnival_sample_result <- file.path(BASE_DIR, paste0(sample_ids, "_carn_res.RDS.gz"))
  names(carnival_sample_result) <- sample_ids
  base_path <- "/desktop-home/heyer/projects/Vascular_Aging"
  tt_basepath <- file.path(base_path, "RNAseq/rna-seq-star-deseq2/workflow/scripts/transcriptutorial")
  organism <- "Mus musculus"
}
source(file.path(tt_basepath, "support_enrichment.r"))
source(file.path(tt_basepath, "support_networks.r"))

carnival_sample_resolution <- purrr::map(carnival_sample_result, readRDS)

# Check if any are 0

length_index <- purrr::map(carnival_sample_resolution, length)
# Check if any are null
lgl_index <- purrr::map_int(carnival_sample_resolution, length) == 0
carnival_sample_resolution <- carnival_sample_resolution[!lgl_index]
stopifnot(length(carnival_sample_resolution) > 0)
process_carnival <- function(carnival_res) {
  carnival_res$weightedSIF <- carnival_res$weightedSIF %>% dplyr::filter(Weight != 0)
  #  carnival_res$nodesAttributes <- carnival_res$nodesAttributes %>% dplyr::filter(AvgAct != 0)
  carnival_res
}

carnival_sample_resolution <- purrr::map(carnival_sample_resolution, process_carnival)
```


WARNING:
The following samples could not create a network.
`r names(carnival_sample_resolution)[length_index == 0]`


```{r}
# read CARNIVAL results
omnipath_id <- RNAscripts::get_organism_omnipath_id(organism)
omniR <- OmnipathR::import_omnipath_interactions(organism = omnipath_id)

# signed and directed
omnipath_sd <- omniR %>% dplyr::filter(consensus_direction == 1 &
  (consensus_stimulation == 1 |
    consensus_inhibition == 1
  ))

# changing 0/1 criteria in consensus_stimulation/inhibition to -1/1
omnipath_sd$consensus_stimulation[which(omnipath_sd$consensus_stimulation == 0)] <- -1
omnipath_sd$consensus_inhibition[which(omnipath_sd$consensus_inhibition == 1)] <- -1
omnipath_sd$consensus_inhibition[which(omnipath_sd$consensus_inhibition == 0)] <- 1

# check consistency on consensus sign and select only those in a SIF format
sif <- omnipath_sd[, c("source_genesymbol", "consensus_stimulation", "consensus_inhibition", "target_genesymbol")] %>%
  dplyr::filter(consensus_stimulation == consensus_inhibition) %>%
  unique.data.frame()

sif$consensus_stimulation <- NULL
colnames(sif) <- c("source", "interaction", "target")

# remove complexes
sif$source <- gsub(":", "_", sif$source)
sif$target <- gsub(":", "_", sif$target)
```

```{r}
# get only summary files from CARNIVAL results
sifts <- lapply(carnival_sample_resolution, function(x) {
  x$weightedSIF
})
nodos <- lapply(carnival_sample_resolution, function(x) {
  x$nodesAttributes
})
write(names(sifts), file = stderr())
# Calculate the number of edges and nodes in the networks and its density
node_edge <- do.call(rbind, lapply(sifts, count_edges_nodes_degree))


# Calculate degree distribution for a sample
yeet <- do.call(rbind, sifts) %>% unique()
count_degree <- yeet %>% degree_count()

# degree distribution
p <- data.frame(table(count_degree$total_count) / nrow(count_degree))
colnames(p) <- c("Var1", "total_degree")
p <- merge.data.frame(p, data.frame(table(count_degree$in_count) / nrow(count_degree)), all = T)
colnames(p) <- c("Var1", "total_degree", "in_degree")
p <- merge.data.frame(p, data.frame(table(count_degree$out_count) / nrow(count_degree)), all = T)
colnames(p) <- c("k", "total_degree", "in_degree", "out_degree")
p <- melt(p, value.name = "p", id.vars = "k")
p$k <- relevel(p$k, "0")

# visualise
ggdat <- as.data.frame(node_edge) %>%
  tibble::rownames_to_column(var = "sample") %>%
  dplyr::mutate(condition = gsub(".Rep[0-9]{1}", "", sample))

# Plotting

# relation between number of edges and nodes
ggplot(ggdat, aes(x = nodes, y = edges, color = as.factor(condition))) +
  geom_point() +
  geom_text(
    label = ggdat$sample,
    check_overlap = TRUE,
    vjust = 0,
    nudge_y = 0.5,
    show.legend = F
  ) +
  theme_bw(base_size = 15) +
  guides(color = guide_legend(title = "Conditions")) +
  ggtitle("Node-edge composition")
```

```{r, fig.width=12, fig.height=8}
# degree distribution
levels(p$k) <- levels(p$k) %>%
  as.numeric() %>%
  sort()
dd <- ggplot(data = p, aes(x = k, y = p, group = variable, color = variable)) +
  geom_point() +
  geom_line() +
  theme_bw() +
  theme(legend.position = "none") +
  guides(color = guide_legend(title = "degree type")) +
  ggtitle("Degree distribution")

ddp <- ggplot(data = p, aes(x = as.numeric(k), y = p, group = variable, color = variable)) +
  geom_point() +
  geom_line() +
  scale_x_continuous(
    breaks = as.numeric(p$k),
    trans = scales::log_trans()
  ) +
  scale_y_log10(
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))
  ) +
  annotation_logticks() +
  theme_bw() +
  guides(color = guide_legend(title = "degree type")) +
  ggtitle("Degree distribution (log scale)") +
  xlab("k (ln)") +
  ylab("p (log10)")

plot_grid(dd, ddp, labels = "auto", rel_widths = c(1, 2))
```
# Enrichment Analysis

We define two different gene sets in order to conduct the enrichment analysis.
The first set contains the nodes that appear in the CARNIVAL output,
and are therefore relevant in the context of our input transcriptomic data. 
The second set contains all the genes in our prior knowledge network which are used as the background.

We also downloaded from **MSigDB https://www.gsea-msigdb.org/ the following dataset: c2.cp.v7.1.symbols.gmt**. 
It contains several pathways from different resources and the genes that are known to be involved in those pathways.

Using the *Piano* R package, 
we run a gene set analysis (GSA) based on a list of significant genes (CARNIVAL nodes) and a gene set collection (background). 
It uses Fisherâ€™s exact test.

```{r enrichment, message=FALSE, results='hide', warning=FALSE}
```

# Topological measurments

We can get a sense of the size of the reconstructed networks using topological parameters, such as number of edges and nodes.
We can also have a look at the network's density and the *degree distribution*.
The *density* indicates the proportion of interactions that exist in our network when comparing with all possible interactions that can be established. 
The *degree distribution* shows the number of connections of a node. 
In a directed network, we can distinguish between incoming and outgoing connections.

NOTE: Here the density is calculated for a _directed graph_.
As CARNIVAL can report 2 interactions between the same 2 nodes with different sign, 
these "doubled" interactions are excluded when calculating the density.

For this section, and the following ones, we use the CARNIVAL networks obtained
when running the _sample_resolution_carnival.R_. 

To find out more about this topic:

+ https://mathinsight.org/degree_distribution
+ https://www.networksciencebook.com/chapter/2#degree

```{r netopology, message=FALSE, results='hide', warning=FALSE}
get_common_interactions <- function(interaction_list, psmpl_per = 95,
                                    s_table = sample_table,
                                    condition_sel) {
  sample_id_list <- s_table %>%
    filter(condition %in% condition_sel) %>%
    pull(sample)
  interaction_list <- interaction_list[, sample_id_list]

  if (!is.null(dim(interaction_list))) {
    shared_interactions_WT <- getCoreInteractions(topology = interaction_list, psmpl = psmpl_per)

    # Visualise the interactions
    colnames(shared_interactions_WT) <- c("from", "sign", "to")
    labels_edge <- c("-1" = "inhibition", "1" = "activation")
    nodes <- data.frame(union(shared_interactions_WT$from, shared_interactions_WT$to))
    colnames(nodes) <- "nodes"
    nodes$label <- nodes$nodes

    gg_graph <- tidygraph::tbl_graph(nodes = nodes, edges = shared_interactions_WT) %>%
      ggraph::ggraph(layout = "auto") +
      geom_node_point(color = "#C0C0C0", size = 8) +
      geom_edge_link(arrow = arrow(), aes(edge_colour = as.factor(sign))) +
      theme_graph() +
      geom_node_text(aes(label = label), vjust = 0.4)
  } else {
    gg_graph <- NULL
  }
  gg_graph
}
```

Generally speaking, biological networks are not dense, so don't worry if the density values are low... they usually are!

The degree distribution is an interesting graphic to look at, as we can get some immediate information:

+ k = 0 for in-degree indicates the proportion of initial nodes, 
while for out-degree indicates the effectors (here the TFs). 
For total-degree, that would indicate the presence of isolated nodes.

+ Most of the nodes do not have many connections (e.g. k = 1 -> p = 0.6), 
but there are a few that are highly connected (e.g. k > 6).

+ We can find some hubs when k is higher (plot B). The out-degree ends at k = 5; 
this means that the biggest regulatory-hub regulates at most 5 other nodes.
In a similar way, the in-degree goes up to k = 9;
This means that there are few hubs (k > 5) that are regulated by at most 9 nodes.

# Network comparison

When we have more than one network, we usually would like to know how (dis)similar these networks are.

We can use the *Jaccard Index* to measure similarities and diversity
between sample sets (https://en.wikipedia.org/wiki/Jaccard_index).
This index can be used in different scenarios.
Generally, it's used with the nodes, but we can also use it with the edges.

When we have groups, in our case WT and FOXA2KO, we can also extract the interactions that are common.

```{r}
# create a matrix of all interactions for all samples
write(dim(sif), file = stderr())
interactions <- getTopology(networks = sifts, scafoldNET = sif)
colnames(interactions) <- names(carnival_sample_resolution)
# FIxes bug in topology function (To lazy to actually fix)
ncol_interact <- ncol(interactions)
# interactions <- interactions[rowSums(interactions) > 0,]
# interactions <- interactions[,-c(1:ncol_interact/2)]
# get the edges per sample
# get the edges per sample
net_int <- apply(interactions, 2, function(x, r) {
  r[which(!is.na(x))]
}, rownames(interactions))

# calculate Jaccard indexes per pair
combined <- expand.grid(1:length(names(sifts)), 1:length(names(sifts)))
jac_index <- matrix(
  data = NA, nrow = length(names(sifts)), ncol = length(names(sifts)),
  dimnames = list(names(sifts), names(sifts))
)

for (i in 1:nrow(combined)) {
  n <- names(sifts)[combined[i, 1]]
  m <- names(sifts)[combined[i, 2]]
  jac_index[n, m] <- length(intersect(net_int[[n]], net_int[[m]])) / length(union(net_int[[n]], net_int[[m]]))
}

# Visualize the indexes in a heatmap

pheatmap::pheatmap(jac_index,
  fontsize = 14,
  fontsize_row = 10, fontsize_col = 10,
  angle_col = 45, treeheight_col = 0
)
corrplot::corrplot(jac_index, order = "hclust")
```


## Inferring mixed networks from samples

Here we investiagte which edges are shared between samples to get a bettter look at the structure of these. networks

We plot Edges which we can find in at least 60% of samples in one comparison group. 

We compare both in one sample group, but we also have a look which edges are consisten across groups (noted with Combined)
aka. combined_young == young-ec + young-EC_apeliun

```{r, fig.width= 8, fig.height=8}
groups_to_check <- c(
  workflow_config$comp_groups,
  as.list(setNames(
    object = sample_table$condition %>% unique(),
    nm = sample_table$condition %>% unique()
  ))
)


yoted <- purrr::map(groups_to_check, get_common_interactions, interaction_list = interactions, psmpl_per = 60, s_table = sample_table[!lgl_index, ])
if (any(purrr::map(yoted, is.null))) {
  null_index <- purrr::map_lgl(yoted, is.null)
  yoted <- yoted[!null_index]
}
yoted <- purrr::map2(yoted, names(yoted), function(x, y) x + ggtitle(y))

yoted
```

```{r, fig.height = 12}
build_matrix_from_nodes <- function(node_list, node_type = "AvgAct") {
  gene_index <- purrr::map(node_list, ~ pull(., Node)) %>%
    unlist() %>%
    unique()
  node_mat <- purrr::map(node_list, ~ dplyr::select(., Node, !!node_type)) %>% purrr::reduce(full_join, by = "Node")
  colnames(node_mat) <- c("Node", names(node_list))
  node_mat
}

avg_mat <- build_matrix_from_nodes(nodos)
rownames(avg_mat) <- avg_mat$Node
avg_mat <- subset(avg_mat, select = -c(Node)) %>% as.matrix()
non_zero_index <- apply(avg_mat, 1, function(x) length(which(x != 0)) >= 2)



ComplexHeatmap::Heatmap(avg_mat[non_zero_index, ], column_names_rot = 45, row_names_gp = grid::gpar(fontsize = 6), cluster_columns = FALSE)
```
## Session Info Details

```{r sessionInfo, echo=FALSE, eval=TRUE}
sessionInfo()
```
